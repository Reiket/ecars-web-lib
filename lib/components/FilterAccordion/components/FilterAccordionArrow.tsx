import type {FC} from 'react';
import {Icons} from '@/services/icons';

interface Props {
  onClick: () => void;
}

export const FilterAccordionArrow: FC<Props> = ({onClick}) => (
  <button
    onClick={onClick}
    className="filter-accordion__arrow"
  >
    <Icons.ArrowNarrowDown />
  </button>
);
