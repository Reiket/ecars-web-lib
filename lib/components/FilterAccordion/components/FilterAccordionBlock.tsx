import type {FC, ReactNode} from 'react';

export interface Props {
  children: ReactNode;
}

export const FilterAccordionBlock: FC<Props> = ({children}) => (
  <div className="filter-accordion__block">{children}</div>
);
