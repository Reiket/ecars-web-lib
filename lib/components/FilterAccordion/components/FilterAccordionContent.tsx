import type {FC, ReactNode} from 'react';
import {useAccordionHeight} from '@/services/hooks/useAccordionHeight';

interface Props {
  isOpen: boolean;
  children: ReactNode;
}

export const FilterAccordionContent: FC<Props> = ({isOpen, children}) => {
  const {style, contentRef} = useAccordionHeight(isOpen);
  return (
    <div
      style={style}
      ref={contentRef}
      className="filter-accordion__content"
    >
      <div className="filter-accordion__body">{children}</div>
    </div>
  );
};
