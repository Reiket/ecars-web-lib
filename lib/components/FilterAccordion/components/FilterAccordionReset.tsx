import type {FC} from 'react';
import {Button} from '@/components/Button/Button';
import type {ElementProps} from '@/services/types';

export const FilterAccordionReset: FC<ElementProps> = ({block}) => (
  <Button
    block={block}
    isTransparent
    size="small"
    color="gray"
  >
    Reset
  </Button>
);
