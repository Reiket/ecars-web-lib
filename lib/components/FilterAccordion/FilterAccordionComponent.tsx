import type {FC, ReactNode} from 'react';
import type {ElementProps} from '@/services/types';
import {FILTER_ACCORDION_BLOCK_CLASS, FilterAccordion} from '@/components/FilterAccordion/constants';

export interface Props extends ElementProps {
  isOpen: boolean;
  title: string;
  onClick: () => void;
  children: ReactNode;
}

export const FilterAccordionComponent: FC<Props> = ({onClick, title, isOpen, children}) => {
  return (
    <FilterAccordion.Layout isOpen={isOpen}>
      <FilterAccordion.Block block={FILTER_ACCORDION_BLOCK_CLASS}>
        <h2 className="filter-accordion__title">{title}</h2>
        <FilterAccordion.Reset />
        <FilterAccordion.Arrow onClick={onClick} />
      </FilterAccordion.Block>
      <FilterAccordion.Content isOpen={isOpen}>{children}</FilterAccordion.Content>
    </FilterAccordion.Layout>
  );
};
