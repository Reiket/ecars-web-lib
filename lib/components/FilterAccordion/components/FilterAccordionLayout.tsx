import type {FC, ReactNode} from 'react';
import {cn} from '@/services/helpers';
import type {ElementProps} from '@/services/types';
import {FILTER_ACCORDION_TEST_ID} from '@/components/FilterAccordion/constants';

interface Props extends ElementProps {
  isOpen: boolean;
  children: ReactNode;
}

export const FilterAccordionLayout: FC<Props> = ({isOpen, block, children}) => (
  <div
    data-testid={FILTER_ACCORDION_TEST_ID}
    className={cn(block, 'filter-accordion', {
      '_active-accordion': isOpen,
    })}
  >
    {children}
  </div>
);
