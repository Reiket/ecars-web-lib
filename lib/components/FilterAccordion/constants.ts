import {FilterAccordionLayout} from '@/components/FilterAccordion/components/FilterAccordionLayout';
import {FilterAccordionArrow} from '@/components/FilterAccordion/components/FilterAccordionArrow';
import {FilterAccordionReset} from '@/components/FilterAccordion/components/FilterAccordionReset';
import {FilterAccordionBlock} from '@/components/FilterAccordion/components/FilterAccordionBlock';
import {FilterAccordionContent} from '@/components/FilterAccordion/components/FilterAccordionContent';
import {withBlockClass} from '@/services/hoc/withBlockClass';
import {FilterAccordionComponent} from '@/components/FilterAccordion/FilterAccordionComponent';

export const FILTER_ACCORDION_BLOCK_CLASS = 'filter-accordion';
export const FILTER_ACCORDION_TEST_ID = 'filterAccordionTestId';
export const FilterAccordionHOC = withBlockClass(FilterAccordionBlock, FILTER_ACCORDION_BLOCK_CLASS);

export const FilterAccordion = Object.assign(FilterAccordionComponent, {
  Layout: FilterAccordionLayout,
  Arrow: FilterAccordionArrow,
  Reset: FilterAccordionReset,
  Block: FilterAccordionHOC,
  Content: FilterAccordionContent,
});
