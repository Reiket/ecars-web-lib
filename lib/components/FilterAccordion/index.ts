import {FilterAccordionComponent} from '@/components/FilterAccordion/FilterAccordionComponent';
import {FilterAccordionLayout} from '@/components/FilterAccordion/components/FilterAccordionLayout';
import {FilterAccordionArrow} from './components/FilterAccordionArrow';
import {FilterAccordionReset} from '@/components/FilterAccordion/components/FilterAccordionReset';
import {FilterAccordionHOC} from '@/components/FilterAccordion/constants';
import {FilterAccordionContent} from './components/FilterAccordionContent';

export const FilterAccordion = Object.assign(FilterAccordionComponent, {
  Layout: FilterAccordionLayout,
  Arrow: FilterAccordionArrow,
  Reset: FilterAccordionReset,
  Block: FilterAccordionHOC,
  Content: FilterAccordionContent,
});
