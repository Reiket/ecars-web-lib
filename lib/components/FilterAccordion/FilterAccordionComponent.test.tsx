import {fireEvent, render, screen} from '@testing-library/react';
import {FILTER_ACCORDION_TEST_ID} from '@/components/FilterAccordion/constants';
import {FilterAccordion} from '@/components/FilterAccordion';

const defaultProps = {
  isOpen: false,
  children: 'Test content',
  title: 'Test title',
  onClick: vi.fn(),
};
describe('Filter Accordion Component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });
  const isOpenValues = [true, false];
  const renderAccordion = (props = {}) => {
    const mergedProps = {...defaultProps, ...props};
    return render(<FilterAccordion {...mergedProps} />);
  };

  test('renders the filter accordion component and title, text correctly', () => {
    const {container} = renderAccordion();
    expect(screen.getByText(defaultProps.title)).toBeInTheDocument();
    expect(screen.getByText(defaultProps.children)).toBeInTheDocument();
    expect(screen.getByTestId(FILTER_ACCORDION_TEST_ID)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  isOpenValues.forEach((isOpen) => {
    const isOpenString = `should ${isOpen ? 'open' : 'close'} the content`;
    test(isOpenString, () => {
      const {container} = renderAccordion({isOpen});
      const content = screen.getByText(defaultProps.children);
      const accordionElement = screen.getByTestId(FILTER_ACCORDION_TEST_ID);
      if (isOpen) {
        expect(content).toBeVisible();
        expect(accordionElement).toHaveClass('_active-accordion');
      } else {
        expect(content.offsetHeight).toBe(0);
        expect(accordionElement).not.toHaveClass('_active-accordion');
        expect(container).toMatchSnapshot();
      }
    });
  });
  test('triggers onClick handler when arrow is clicked', () => {
    renderAccordion();
    const filterAccordionElement = screen.getByTestId(FILTER_ACCORDION_TEST_ID);
    const buttonArrow = filterAccordionElement.querySelector('.filter-accordion__arrow');
    if (buttonArrow) {
      fireEvent.click(buttonArrow);
    }
    expect(defaultProps.onClick).toHaveBeenCalledTimes(1);
  });
});
