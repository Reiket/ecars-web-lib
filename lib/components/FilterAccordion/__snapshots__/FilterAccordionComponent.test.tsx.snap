// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Accordion Component renders the filter accordion component and title, text correctly 1`] = `
<div>
  <div
    class="filter-accordion"
    data-testid="filterAccordionTestId"
  >
    <div
      class="filter-accordion__block"
    >
      <h2
        block="filter-accordion"
        class="filter-accordion__title"
      >
        Test title
      </h2>
      <button
        class="filter-accordion__button button button--small button--gray button--transparent"
        data-testid="buttonTestId"
        type="button"
      >
        Reset
      </button>
      <button
        class="filter-accordion__arrow"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
          />
        </svg>
      </button>
    </div>
    <div
      class="filter-accordion__content"
      style="height: 0px;"
    >
      <div
        class="filter-accordion__body"
      >
        Test content
      </div>
    </div>
  </div>
</div>
`;
